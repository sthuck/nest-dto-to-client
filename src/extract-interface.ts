import {
  ArrayLiteralExpression,
  CallExpression,
  ClassDeclaration,
  Node,
  SourceFile,
  SyntaxKind,
} from 'ts-morph';

export function ExtractInterface(c: ClassDeclaration, outputFile: SourceFile) {
  const i = c.extractInterface();
  const e = c.getExtends();
  if (e) {
    const [extendsStatement] = e.getChildren();

    if (Node.isCallExpression(extendsStatement)) {
      i.extends = [extractInterfaceFromCallStatement(extendsStatement)];
    } else {
      i.extends = [e.getText()];
    }
  }
  i.isExported = true;
  outputFile.addInterface(i);
}

const OmitPickTemplateHelper = (args: Node[]) => {
  const extendedClass = args[0];
  let extendedClassStr = Node.isCallExpression(extendedClass)
    ? extractInterfaceFromCallStatement(extendedClass)
    : extendedClass.getText();

  let keys = args[1];

  if (Node.isAsExpression(keys)) {
    keys = keys.getChildrenOfKind(SyntaxKind.ArrayLiteralExpression)[0];
  }

  if (Node.isArrayLiteralExpression(keys)) {
    const keysJoined = keys
      .getElements()
      .map((n) => n.getText())
      .join('|');
    return [extendedClassStr, keysJoined];
  }
  const argsText = `[${args.map((n) => n.getText()).join(',')}]`;
  throw new Error('No array literal for Omit/Pick type ' + argsText);
};

type NestToNativeTypeConvert = {
  template(args: Node[]): string;
};

const FnMap: Record<string, NestToNativeTypeConvert> = {
  PartialType: {
    template: (args: Node[]) => {
      const child = args[0];
      const content = Node.isCallExpression(child)
        ? extractInterfaceFromCallStatement(child)
        : child.getText();
      return `Partial<${content}>`;
    },
  },
  OmitType: {
    template: (args: Node[]) => {
      const [extendedClassStr, keysJoined] = OmitPickTemplateHelper(args);
      return `Omit<${extendedClassStr}, ${keysJoined}>`;
    },
  },
  PickType: {
    template: (args: Node[]) => {
      const [extendedClassStr, keysJoined] = OmitPickTemplateHelper(args);
      return `Pick<${extendedClassStr}, ${keysJoined}>`;
    },
  },
};

function extractInterfaceFromCallStatement(extendsStatement: CallExpression) {
  const [fnName, ...args] = extendsStatement.forEachChildAsArray();

  const typeConvert = FnMap[fnName.getText()];
  return typeConvert.template(args);
}
