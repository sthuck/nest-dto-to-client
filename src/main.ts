import { Project } from 'ts-morph';
import { ExtractInterface } from './extract-interface';
import * as path from 'path';

export const main = async (tsConfigPath) => {
  const project = new Project({ tsConfigFilePath: tsConfigPath });
  const files = project.getSourceFiles();
  const dtos = files.filter((f) => f.getBaseName().includes('.dto.ts'));
  const newProject = new Project({ useInMemoryFileSystem: true });

  for (const dtoFile of dtos) {
    const name = dtoFile.getBaseName();
    const outputFile = newProject.createSourceFile(name);

    const imports = dtoFile.getImportDeclarations();
    imports.forEach((i) => {
      const importedFrom = i.getModuleSpecifier().getLiteralValue();
      if (importedFrom.includes('.dto')) {
        const importStructure = i.getStructure();
        importStructure.moduleSpecifier =
          '.' + path.sep + path.basename(importStructure.moduleSpecifier);
        outputFile.addImportDeclaration(importStructure);
      }
    });

    await Promise.all(
      dtoFile.getClasses().map(async (c) => {
        ExtractInterface(c, outputFile);
      }),
    );
  }

  const result = newProject.getSourceFiles().map((file) => {
    const text = file.getFullText();
    const name = file.getBaseName();
    return { text, name };
  });

  return result;
};
