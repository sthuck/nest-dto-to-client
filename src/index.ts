import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { main } from './main';
import * as fs from 'fs';
import * as path from 'path';

const driver = async () => {
  const argv = yargs(hideBin(process.argv))
    .options({
      o: {
        demandOption: true,
        description: 'output dir',
        type: 'string',
      },
      p: {
        demandOption: true,
        description: 'tsconfig location',
        type: 'string',
      },
    })
    .help('h')
    .parseSync();

  const outputDir = argv['o'];
  const projectFile = argv['p'];

  const outputDirStat = fs.statSync(outputDir);
  if (!outputDirStat.isDirectory()) {
    throw new Error(`${outputDir} not a valid directory`);
  }

  const tsconfigState = fs.statSync(projectFile);
  if (!tsconfigState.isFile()) {
    throw new Error(`${projectFile} not a valid file`);
  }

  const generatedFiles = await main(projectFile);
  generatedFiles.forEach((file) => {
    const fullPath = path.resolve(outputDir, file.name);
    fs.writeFileSync(fullPath, file.text, 'utf8');
  });
};

driver().catch(console.error);
