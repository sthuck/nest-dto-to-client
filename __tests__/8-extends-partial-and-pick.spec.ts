import { main } from '../src/main';

describe('extends partial types, with nested complex types', () => {
  it('extends partial types, with nested complex types', async () => {
    const result = await main(
      './test/fixtures/8-extends-partial-and-pick/tsconfig.json',
    );
    expect(result).toMatchSnapshot();
  });
});
