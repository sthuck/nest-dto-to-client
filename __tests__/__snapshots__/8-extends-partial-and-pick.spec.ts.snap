// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extends partial types, with nested complex types extends partial types, with nested complex types 1`] = `
Array [
  Object {
    "name": "nested.dto.ts",
    "text": "export interface NestedDto {
    a: number;
    b: string;
    c: boolean;
}
",
  },
  Object {
    "name": "simple.dto.ts",
    "text": "import { NestedDto } from \\"./nested.dto\\";

export interface SimpleDto2 extends Pick<Partial<NestedDto>, 'a'|'c'> {
}

export interface SimpleDto extends Partial<Pick<NestedDto, 'a'|'c'>> {
}
",
  },
]
`;
